internal/go_repository.bzl

```diff
@@ -0,0 +1,236 @@
+      fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
+
+  generate = ctx.attr.build_file_generation == "on"
+  if ctx.attr.build_file_generation == "auto":
+    generate = True
+    for name in ['BUILD', 'BUILD.bazel', ctx.attr.build_file_name]:
+      path = ctx.path(name)
+      if path.exists and not env_execute(ctx, ['test', '-f', path]).return_code:
+        generate = False
+        break
+  if generate:
+    # Build file generation is needed
+    _gazelle = "@bazel_gazelle_go_repository_tools//:bin/gazelle{}".format(executable_extension(ctx))
+    gazelle = ctx.path(Label(_gazelle))
+    cmd = [gazelle, '--go_prefix', ctx.attr.importpath, '--mode', 'fix',
+            '--repo_root', ctx.path('')]
+    if ctx.attr.build_file_name:
+      cmd.extend(["--build_file_name", ctx.attr.build_file_name])
+    if ctx.attr.build_tags:
+      cmd.extend(["--build_tags", ",".join(ctx.attr.build_tags)])
+    if ctx.attr.build_external:
+      cmd.extend(["--external", ctx.attr.build_external])
+    if ctx.attr.build_file_proto_mode:
+      cmd.extend(["--proto", ctx.attr.build_file_proto_mode])
+    cmd.append(ctx.path(''))
+    result = env_execute(ctx, cmd)
+    if result.return_code:
+      fail("failed to generate BUILD files for %s: %s" % (
+          ctx.attr.importpath, result.stderr))
+
+go_repository = repository_rule(
+    implementation = _go_repository_impl,
+    attrs = {
+        # Fundamental attributes of a go repository
+        "importpath": attr.string(mandatory = True),
+
+        # Attributes for a repository that should be checked out from VCS
+        "commit": attr.string(),
+        "tag": attr.string(),
+        "vcs": attr.string(
+            default = "",
+            values = [
+                "",
+                "git",
+                "hg",
+                "svn",
+                "bzr",
+            ],
+        ),
+        "remote": attr.string(),
+
+        # Attributes for a repository that comes from a source blob not a vcs
+        "urls": attr.string_list(),
+        "strip_prefix": attr.string(),
+        "type": attr.string(),
+        "sha256": attr.string(),
+
+        # Attributes for a repository that needs automatic build file generation
+        "build_external": attr.string(
+            values = [
+                "",
+                "external",
+                "vendored",
+            ],
+        ),
+        "build_file_name": attr.string(default = "BUILD.bazel,BUILD"),
+        "build_file_generation": attr.string(
+            default = "auto",
+            values = [
+                "on",
+                "auto",
+                "off",
+            ],
+        ),
+        "build_tags": attr.string_list(),
+        "build_file_proto_mode": attr.string(
+            values = [
+                "",
+                "default",
+                "disable",
+                "legacy",
+            ],
+        ),
+    },
+)
+"""See go/workspace.rst#go-repository for full documentation."""
```

Comment:
Probably a lie?

Reply:
I just created a new placeholder file to link to. I'll fill it in with documentation about all three repository rules in a follow-up PR.